library/network/src/modules/SuSEFirewall.rb

```diff
@@ -687,6 +688,227 @@ def IsSupportedProtocol(protocol)
       Builtins.contains(@supported_protocols, protocol)
     end
 
+    # Function sets additional ports/services from taken list. Firstly, all additional services
+    # are removed also with their aliases. Secondly new ports/protocols are added.
+    # It uses GetAdditionalServices() function to get the current state and
+    # then it removes what has been removed and adds what has been added.
+    #
+    # @param [String] protocol
+    # @param [String] zone
+    # @param	list <string> list of ports/protocols
+    # @see #GetAdditionalServices()
+    #
+    # @example
+    #	SetAdditionalServices ("TCP", "EXT", ["53", "128"])
+    def SetAdditionalServices(protocol, zone, new_list_services)
+      new_list_services = deep_copy(new_list_services)
+      old_list_services = Builtins.toset(GetAdditionalServices(protocol, zone))
+      new_list_services = Builtins.toset(new_list_services)
+
+      if new_list_services != old_list_services
+        SetModified()
+
+        add_services = []
+        remove_services = []
+
+        # Add these services
+        Builtins.foreach(new_list_services) do |service|
+          if !Builtins.contains(old_list_services, service)
+            add_services = Builtins.add(add_services, service)
+          end
+        end
+        # Remove these services
+        Builtins.foreach(old_list_services) do |service|
+          if !Builtins.contains(new_list_services, service)
+            remove_services = Builtins.add(remove_services, service)
+          end
+        end
+
+        if Ops.greater_than(Builtins.size(remove_services), 0)
+          Builtins.y2milestone(
+            "Removing additional services %1/%2 from zone %3",
+            remove_services,
+            protocol,
+            zone
+          )
+          RemoveAllowedPortsOrServices(remove_services, protocol, zone, true)
+        end
+        if Ops.greater_than(Builtins.size(add_services), 0)
+          Builtins.y2milestone(
+            "Adding additional services %1/%2 into zone %3",
+            add_services,
+            protocol,
+            zone
+          )
+          AddAllowedPortsOrServices(add_services, protocol, zone)
+        end
+      end
+
+      nil
+    end
+
+    # Local function removes ports and their aliases (if check_for_aliases is true), for
+    # requested protocol and zone.
+    #
+    # @param	list <string> ports to be removed
+    # @param [String] protocol
+    # @param [String] zone
+    # @param	boolean check for port-aliases
+    def RemoveAllowedPortsOrServices(remove_ports, protocol, zone, check_for_aliases)
+      remove_ports = deep_copy(remove_ports)
+      if Ops.less_than(Builtins.size(remove_ports), 1)
+        Builtins.y2warning(
+          "Undefined list of %1 services/ports for service",
+          protocol
+        )
+        return
+      end
+
+      SetModified()
+
+      # FirewallD is easy
+      if self.is_a?(SuSEFirewalld)
```

Comment:
This should be a specialized `SuSEFirewalld#RemoveAllowedPortsOrServices` method instead.